### ISSUE TITLE
[Accessibility] Editor Tabs and Source Tabs should be accessible

### ISSUE THREAD

**Issue by yzen** [firefox-devtools/debugger#4075]
There are several improvements that can be done for the source tabs to make them awesome for accessibility:

- [ ] Most Importantly both tab widgets (source and editor tabs) should be the same shared component component that would be styled/configured differently for a particular location.
- [ ] Reachable by keyboard: tabs should be accessible with the keyboard, right now there's not way to focus on the tab (if there is, it is not visible)
- [ ] Tabs should be navigatable via keyboard -- there's a blueprint for adding this functionality here;  https://github.com/devtools-html/debugger.html/pull/5806
- [ ] Styling for keyboard focus for tabs
- [ ] Close tab and add new (+) should be real buttons (or with role button), alternative text and be keyboard accessible: tabbable, activatable and have focus styling
- [x] source-tabs should semantically be tabs. Marco has a good [article] about this.

[article]:https://www.marcozehe.de/2013/02/02/advanced-aria-tip-1-tabs-in-web-apps/

### Editor Tabs

![Screen Shot 2017-09-18 at 8.13.49 PM.png](https://shipusercontent.com/328ae611c2f13327e81689f4bd377f9a/Screen%20Shot%202017-09-18%20at%208.13.49%20PM.png)

### Source Tabs

<img width="254" alt="screen shot 2018-04-27 at 10 33 09" src="https://user-images.githubusercontent.com/561275/39368693-453c6442-4a08-11e8-9235-a5b97e09f062.png">

**Comment by jasonLaster** [comment]
@garbee looked at our tabs before and found them difficult to make accessible given the HTML structure. Does it make sense to try and refactor the HTML or do some JS controls around where tab sound take you?

**Comment by Garbee** [comment]
Styling would be handled the same as buttons in most cases. As that is essentially what tabs are at the end of it all.

The specific problem in carrying this out with tabs though, is the close trigger. Those should be nested within the tab node, and buttons can't be nested in buttons. That's not nice. One solution to this though is borrowing from the linked tutorial. Have the list items, with anchors for the triggers and the close buttons as a button beside the anchor in the markup. This would allow for the anchor to be properly accessible and stylable, while allowing a button node to handle the close trigger which is what makes the most sense there. Win-Win on both sides. All the triggers are fully accessible and easily stylable.

<figure>

```html
<ul> <!-- Because tabs are lists -->
  <li>
    <a href>app-view.js</a><button>close</button>
  <li>
    <a href>app.js</a><button>close</button>
</ul>
```

<figcaption>
  Example markup structure for accessible tabs without all of the accessibility specific attributes.
</figcaption>

</figure>

**Comment by jasonLaster** [comment]
#### should be keyboard navigable

* left and right arrow should change the focused tab
* space or enter should select a source
* tab/shift tab should jump to the command bar or source tree

We'll listen for arrows and space bar events and set the `ariaActiveDescendent` state

```diff
--- a/src/components/Editor/Tabs.js
+++ b/src/components/Editor/Tabs.js
@@ -325,6 +325,17 @@ class SourceTabs extends PureComponent {
     );
   }

+  onKeyDown() {
+    if (arrow) {
+      return this.setState({ariaActiveDescendent: tabIndes})
+    }
+
+    if (spaceBar) {
+      const source = getSource(tabIndex)
+      this.props.selectSource(source);
+    }
+  }
```

**Comment by CJSt** [comment]
/claim

Our team will work on this as part of an assignment for CSC426.

**Comment by CJSt** [comment]
Add new button has been removed in another update, so there's not a need to look for that. We weren't able to convert the tabs into lists in order to get keyboard navigation to work properly. Also, this issue seems similar to issue #4070 where the close button is also needing keyboard navigation.

**Comment by jasonLaster** [comment]
@yzen @darkwing what would you like to do here?

**Comment by darkwing** [comment]
We'll need to do a larger review and make some decisions.  Making the close button an actual button seems reasonably easy.  We'll need to think about keyboard shortcuts and tabbing, however.  I'm thinking `CTRL+0`, CTRO+1`, etc. would work for selecting individual tabs.  How we navigate from one to the next....that will take some thought.

**Comment by darkwing** [comment]
I drafted keyboard navigation in this pull request:  https://github.com/devtools-html/debugger.html/pull/5806

I'd love for someone to drag it across the finish line!

**Comment by Sekhmet** [comment]
/claim

I'd love to work on that, but it might take a while as I'm on a tight schedule.
If someone is willing to work on that as well and can finish it faster let me know.

**Comment by rajatvijay** [comment]
/claim

Would love to work on it. 
Might take some time to set it up and understand the current code. But will try to finish as soon as I can.

**Comment by ahuth** [comment]
/claim

Figured I'll take a shot at this, as I have experience implementing [tabs per the WAI-ARIA authoring practices 1.1](https://www.w3.org/TR/wai-aria-practices-1.1/#tabpanel) for a [React accessible component library](https://github.com/juanca/react-aria-components).

Worst case I'll unclaim if it ends up being more than I can handle, and will at least report back with any findings.

**Comment by jasonLaster** [comment]
Hey @ahuth what are your thoughts on the editor tabs? feasible?

**Comment by ahuth** [comment]
The current UX is definitely feasible. The hard part I think will be that the implementation needs to be changed pretty significantly for them.

**Comment by jasonLaster** [comment]
got it

**Comment by fvsch** [comment]
/claim

I'd like to work on this.

About the "left/right should change the current tab", that means [implementing tabs as a menu role with ARIA and a lot of JS](http://adrianroselli.com/2017/10/dont-use-aria-menu-roles-for-site-nav.html), which may not be compatible with having buttons for the close action inside or alongside each `menuitem`. So maybe just having a <nav> with a <ul> list and two buttons in each item (one for enabling a tab, one for closing) would work better.

I think we may want to handle the sources/outline tab implementation separately too, especially since we discussed moving away from tabs and showing the outline below the sources in the left sidebar (as a collapsible pane) at All Hands.

**Comment by ahuth** [comment]
I'm not sure it necessarily requires using a menu role, @fvsch. For example, the "outline" tabs use [react-aria-components](https://github.com/juanca/react-aria-components), which implement the [WAI-ARIA Authoring Practice's Tabs](https://www.w3.org/TR/wai-aria-practices-1.1/#tabpanel).

We may be better being consistent and staying with the authoring practice's tabs, which use arrows to move left and right. There's also already a react component in the project which is capable of meeting that.

Having said that, I completely failed to finish this issue out and implement tabs for the editor tabs 😿 , because they're implemented in a (very reasonable) way that unfortunately conflicted with how the react-aria-component tabs work, and would require some refactoring.

So please take a shot at coming up with a solution!

**Comment by janodvarko** [comment]
Firefox Developer Tools (including Debugger tool) development is now entirely happening in Bugzilla and new issues should be reported here: https://bugzilla.mozilla.org/enter_bug.cgi?product=DevTools&component=Debugger

Please reopen this report in Bugzilla if it's still valid.

Honza
